Expose ExtractItems as the opposite of RemoveItems

kevindelgado opened the pull request:
This exposes a new method on TypedValue called ExtractItems. Whereas
RemoveItems takes a fieldpath.Set adn returns the typed value with the
provided set items removed from it, ExtractItems returns ONLY those
items provided by the fieldpath.Set, effectively extracting them from
the original TypedValue.

kevindelgado commented:
/assign @apelisse 
/cc @Jefftree @jpbetz

kevindelgado commented:
As we discussed, the changes to modify remove into extract don’t seem to justify needing an entirely separate walker. That being said, I’m not a fan of the naming. Here I’ve left  it as `removeWalker` with a `shouldExtract` flag.  I considered calling it `extractOrRemoveWalter`, but didn’t know if that was better or worse. Some other thoughts I had were, `isolatingWalker`, `separatingWalker`, etc. Definitely open to any suggestions.

kevindelgado commented:
For testing I mostly pattern-matched off of what seemed relevant from `merge_test` and `toset_test`.  I feel like I might be missing some, so let me know if any edge cases aren’t being tested or alternatively if any seem redundant and should be removed. 

The copypasta of shared testing schemas feels a little wrong. Let me know if we want to organize them better in some sort of shared util file or something (I can do it in a separate PR since it’s not really relevant here).

I attempted to test with a more nested schema such as https://github.com/kubernetes-sigs/structured-merge-diff/blob/master/merge/nested_test.go#L28-L112 but I ran into some difficulties. Maybe I was just constructing the typed values and fieldsets wrong, but it seemed like whatever fieldset I passed it would always remove nothing (leaving the original value) and extract nothing (leaving nil/empty value). Here’s an example of what I’m talking about for the schema above

```
                `{"listOfLists":[{"name": "a", "value":["b", "c"]}, {"name": "d"}]}`, // input typed value
                _NS(_P("listsOfLists", _KBF("name", "d"))),                                      // input fieldset to remove/extract
                `{"listOfLists":[{"name": "a", "value":["b", "c"]}]}`,                          // expected remove output
                `{"listOfLists":[{"name": "d"}]}`,                                                        // expected extract output
```
and got
```
--- FAIL: TestRemove (0.00s)
    --- FAIL: TestRemove/nested_types (0.00s)
        --- FAIL: TestRemove/nested_types/nested_types-valid-1 (0.00s)
            remove_test.go:449: Expected
                listOfLists=[name="a"value=["b","c"]]
                but got
                listOfLists=[name="a"value=["b","c"],name="d"]
            remove_test.go:464: Expected
                listOfLists=[name="d"]
                but got
                null

``` 
I noticed that none of the tests in the typed package use nested schemas (only tests in the merge pkg do). So I’m guessing there’s something I don’t know about nested types that makes unit testing them in the typed package not correct. Do you know?

apelisse commented on typed/remove_test.go:
```diff
@@ -0,0 +1,280 @@
+			}
+
+			set := quadruplet.set
+			if err != nil {
```
There's no new error here :-) Also not sure you need the `set` variable.

kevindelgado replied:
whoops, old code. fixed

apelisse commented on typed/remove.go:
```diff
@@ -20,19 +20,26 @@ import (
+// of the input value with either:
+// 1. only the items in the toRemove set (when shouldExtract is true) or
+// 2. the items from the toRemove set removed from the value (when shouldExtract is false).
+func removeItemsWithSchema(val value.Value, toRemove *fieldpath.Set, schema *schema.Schema, typeRef schema.TypeRef, shouldExtract bool) value.Value {
```
At first I was curious if we wanted to rename this type/method since it's not doing the same thing now, but it's still actually removing items. So i think we're good.

kevindelgado replied:
👍

apelisse commented on typed/remove_test.go:
```diff
@@ -0,0 +1,280 @@
+      scalar: string
+`
+
+var removeCases = []removeTestCase{{
```
These tests look good to me.

kevindelgado replied:
👍

apelisse reviewed (commented):
One super minor nit, thanks!

kevindelgado commented:
feedback addressed, PTAL @apelisse (also, don't forget to /ok-to-test it because I cannot).

apelisse commented:
That nested test looks good, we probably need more of these, not sure why it's failing yet.

apelisse commented on typed/remove_test.go:
```diff
@@ -0,0 +1,275 @@
+
+			rmGot := tv.RemoveItems(quadruplet.set)
+			if !value.Equals(rmGot.AsValue(), rmOut.AsValue()) {
+				t.Errorf("Expected\n%v\nbut got\n%v\n",
```
Can we change the string here to differentiate betwee tne "Removed" and "Extracted"?

kevindelgado replied:
done

kevindelgado commented:
Fixed the issue with doList and added all the nested tests I could think of. Tried to be systematic about it so that we have full coverage of the same tests across all nested types (listOfLists, listOfMaps, mapOfLists, mapOfMaps). 

There are some discrepancies between list-as-top-level types and maps-as-top-level types which creates slightly different test results, I've highlighted these in the inline test comments, but I think these discrepancies are mainly due to the fact that the schema defines listOf* types as requiring name and value whereas mapOf* types do not (see the nested types schema). We can change the schema if we want the tests to look more symmetric, but the added listOf* constraints make it clear what the limitations of `ExtractItems` are. 

Let me know if I'm missing any tests or if you think the bahavior is not what we want. 
PTAL @apelisse

apelisse commented:
/ok-to-test

apelisse commented on typed/remove_test.go:
```diff
@@ -0,0 +1,646 @@
+			_P("mapOfMaps", "b", "q"),
+		),
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		`{"mapOfMaps": {"b":null}}`,
```
Should we even keep `b` here? Or should we even keep anything at all since the one thing we're supposed to keep doesn't exist?

kevindelgado replied:
Yea, that's a good question. I guess my thinking for keeping it this way is that the caller can use the fieldpath.Set they passed in to know that the value did not have a path to the element they expected. If we really don't want it to keep b, we could do some post-processing in `ExtractItems` to get rid of it, that way we don't have to modify the walker, because I think doing it in the walker gets tricky.

apelisse commented on typed/remove_test.go:
```diff
@@ -0,0 +1,646 @@
+		// path to root type
+		`{"mapOfMaps": {"b":{"a":"x","c":"z"}, "d":{"e":"y", "f":"w"}}}`,
+		_NS(
+			_P("mapOfMaps"),
```
we need to think about that too, that's not completely obvious to me

kevindelgado replied:
This is exactly how Remove works currently. If you pass remove a fieldset with only the root element, it will remove everything and return an empty value, so I would Extract to work the same way in reverse, where passing a the root element will "extract" (i.e. keep) everything and return a copy of the original value.

apelisse commented:
Considering the goal we have for this function, I think that a good way to test/agree on what the behavior should be could go this way:
1. Create an object with some fields
2. Apply only some fields to that object, look at the resulting fieldset for that applier
3. Extract from the object using the fieldset we got in 2., we should get back what was applied.

I suspect we should set-up a test for that. A good way to do/understand 2.  is to look at "ToFieldSet", I think if we make the behaviors match there, we should be good to go.

apelisse commented:
> 1. Create an object with some fields
> 2. Apply only some fields to that object, look at the resulting fieldset for that applier
> 3. Extract from the object using the fieldset we got in 2., we should get back what was applied.

In order to keep this test as minimal as possible, I would probably use `typed.Merge` to merge, and `typed.ToFieldSet` to get "apply" and "get the fieldset". That way you don't depend on any other package.

Basically:
1. `obj := typed.AsTyped(basedObject)`
2. `obj := obj.Merge(appliedObject)`, and `fieldset := obj.ToFieldSet()`
3. `assert(obj.Extract(fieldset) == appliedObject)`

kevindelgado commented:
Okay, I think I’ve got the reversibleExtract tests, but there’s one little quirk I need you to sign off on:

`ExtractItems` and `ToFieldSet` are reversible, just not directly. Going form `ToFieldSet` to `ExtractItems` requires filtering out non-leaf nodes.

The issue here is that the fieldset returned by `ToFieldSet` contains all the intermediate nodes in the fieldpath.Set (tree). Thus if you try to extract the set directly, it will return everything under the first path in the set.  

As a result, I think both `ToFieldSet` and `ExtractItems` are behaving correctly; they are just not exactly reversible. They are reversible though if you apply a filter on the output of `ToFieldSet` to remove all the intermediate nodes. This is what I have done here in the tests, and I imagine this is how we want to do it in the clients (probably via some helper we provide).

The alternative is to try to extract only the leafs of the provided fieldpath.Set. I believe this would be incorrect behavior, because you would be unable to extract a top level element and receive all its nested elements (which I believe is the desired behavior).

You might be able to find a way to may `ExtractItems` and `ToFieldSet` reversible if you required `ExtractItems` to only accept fieldpath.Sets that contain only leaf nodes, but this is highly inconsistent with how `RemoveItems` behaves and would be super complicated (maybe impossible) with the walker and resolveSchema paradigm. Furthermore, I think we can confidently say that `ExtractItems` and `ToFieldSet` are deterministically reversible, albeit just not as simple as we hoped it would be by directly piping them into each other.

apelisse commented on typed/remove_test.go:
```diff
@@ -0,0 +1,791 @@
+	}},
+}}
+
+func filterNonLeaves(set *fieldpath.Set) *fieldpath.Set {
```
This sounds like a reasonable implementation to have directly on the `fieldpath.Set` object

apelisse commented on fieldpath/set.go:
```diff
@@ -206,6 +206,32 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
 
+// Leaves returns a set containing only the leaf paths
+// of a set.
+func (s *Set) Leaves() *Set {
```
That's probably not optimal, but I would also suggest against spending time optimizing for something we may not need to optimize.

apelisse replied:
On the other hand, the "optimal" solution might be easier to read.

kevindelgado replied:
Attempted a more performant solution.

kevindelgado reviewed (commented):
Addressed feedback and added `Leaves()` as a method on the `fieldpath.Set` object (along with testing).

PTAL @apelisse

apelisse commented on fieldpath/set.go:
```diff
@@ -206,6 +206,32 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
+		}
+		paths = paths[:n]
+		// save a copy of current path or else it can be overwritten
+		pathCopy := Path(make([]PathElement, len(p)))
```
Any reason why you're not inserting directly in the set?

apelisse commented on fieldpath/set.go:
```diff
@@ -206,6 +206,32 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
+	s.Iterate(func(p Path) {
+		parentPath := Path(p[0 : len(p)-1])
+		n := 0
+		for _, path := range paths {
```
isn't this O(n^2) over the number of paths?

apelisse commented on fieldpath/set_test.go:
```diff
@@ -140,6 +140,12 @@ func BenchmarkFieldSet(b *testing.B) {
+		b.Run(fmt.Sprintf("leaves-%v", here.size), func(b *testing.B) {
+			b.ReportAllocs()
+			for i := 0; i < b.N; i++ {
+				randOperand().Leaves()
```
I'm glad we have a benchmark!

apelisse commented on fieldpath/set_test.go:
```diff
@@ -456,6 +462,73 @@ func TestSetIntersectionDifference(t *testing.T) {
+		expected *Set
+	}{
+		{
+			name:     "empty set",
```
Tests look good, thanks!

kevindelgado commented:
Added a simple ExtractApply test suite based on our conversation yesterday.
On second look, I don't think the big, hairy, multiple appliers tests added any value that isn't captured by these tests so I removed them, but I can add them/some of them back if you think there's value. Let me know if you think there are any cases I'm missing. 

Also, I didn't do any APIVersion checking, because it doesn't seem like there's anything unique to ExtractApply that needs to be tested w.r.t. to APIVersion, but I could be misunderstanding what we want out of an APIVersion check. I know you said we should fail if APIVersion is inconsitent, but it didnt' seem like merge_tests is the place to test that. If it's not already being tested somewhere else (like at a higher level), let me know where I should add tests for it.

I'm working on optimizing set.Leaves() now, but wanted to get the tests up for review in the meantime.

PTAL @apelisse

kevindelgado reviewed (commented):
I've attempted a more performant implementation of `fieldpath.Set#Leaves`. It uses the Set's builtin `Children` and `Members` fields as discussed and inserts directly into the Set. I'm open to any suggestions to making it better.

Compared to the old way it is much faster and uses up a bit less space.

old:
```
BenchmarkFieldSet/leaves-20-16                             76585             15869 ns/op            7256 B/op         95 allocs/op
BenchmarkFieldSet/leaves-50-16                             18986             63825 ns/op           22061 B/op        268 allocs/op
BenchmarkFieldSet/leaves-100-16                             6392            209026 ns/op           75540 B/op        828 allocs/op
BenchmarkFieldSet/leaves-500-16                              565           2200738 ns/op          381671 B/op       3949 allocs/op
BenchmarkFieldSet/leaves-1000-16                             171           7137739 ns/op          822994 B/op       8319 allocs/op
```

new:
```
BenchmarkFieldSet/leaves-20-16                            111402             10611 ns/op            4486 B/op         70 allocs/op
BenchmarkFieldSet/leaves-50-16                             31767             38146 ns/op           15083 B/op        213 allocs/op
BenchmarkFieldSet/leaves-100-16                            10000            123891 ns/op           56641 B/op        720 allocs/op
BenchmarkFieldSet/leaves-500-16                             1759            720563 ns/op          284741 B/op       3434 allocs/op
BenchmarkFieldSet/leaves-1000-16                             777           1579806 ns/op          613902 B/op       7302 allocs/op
```

PTAL @apelisse

kevindelgado commented on internal/fixture/state.go:
```diff
@@ -341,6 +341,84 @@ func (f ForceApplyObject) preprocess(parser Parser) (Operation, error) {
+			set = mgr.Set().Leaves()
+		}
+		// ExtractFields from the state object based on the set
+		extracted := state.Live.ExtractItems(set)
```
I think I've done this, now. Take a look and let me know if that's what you're expecting

kevindelgado commented:
Addressed comments relating to `ExtractApply` operation and further optimized the `Leaves` call.
New benchmark is slightly faster:
```
BenchmarkFieldSet/leaves-20-16                            224038              5368 ns/op            3269 B/op         51 allocs/op
BenchmarkFieldSet/leaves-50-16                             61814             19330 ns/op            9574 B/op        164 allocs/op
BenchmarkFieldSet/leaves-100-16                            15668             75564 ns/op           35945 B/op        690 allocs/op
BenchmarkFieldSet/leaves-500-16                             2814            421019 ns/op          172249 B/op       3432 allocs/op
BenchmarkFieldSet/leaves-1000-16                            1197            999801 ns/op          388709 B/op       7713 allocs/op
```

For posterity, an attempt was made to optimize even further by using two pointers to iterate through the members list once instead of using nested for loops. That had an almost identical benchmark profile (actually, ever so slightly slower) and was much less readable.

That attempt lives here https://github.com/kevindelgado/structured-merge-diff/blob/double-pointer-leaves/fieldpath/set.go#L209-L246

apelisse commented on fieldpath/set.go:
```diff
@@ -455,3 +478,17 @@ func (s *SetNodeMap) iteratePrefix(prefix Path, f func(Path)) {
+			pathElement: n.pathElement,
+			set:         n.set.Leaves(),
+		}
+		out.members = append(out.members, next)
```
Since we know how many members we have, I think we can pre-allocate the list beforehand. That probably has more impact that the double loop above :-)

kevindelgado replied:
I've added preallocation both here and with the `PathElementSet` [here](https://github.com/kubernetes-sigs/structured-merge-diff/pull/185/files#diff-cbe655e4df83a27b44de60b4d4008cbcf550462c360f95d36aeb9826a2af525fR212)

makes it ever so slightly faster
```
BenchmarkFieldSet/leaves-20-16                            317458              3785 ns/op            2250 B/op         41 allocs/op
BenchmarkFieldSet/leaves-50-16                             82166             14462 ns/op            6881 B/op        141 allocs/op
BenchmarkFieldSet/leaves-100-16                            20643             60226 ns/op           29160 B/op        632 allocs/op
BenchmarkFieldSet/leaves-500-16                             2984            363610 ns/op          146550 B/op       3212 allocs/op
BenchmarkFieldSet/leaves-1000-16                            1407            851304 ns/op          325102 B/op       7189 allocs/op
```

apelisse commented on fieldpath/set.go:
```diff
@@ -206,6 +206,29 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
+	members := PathElementSet{}
+	for _, member := range s.Members.members {
+		isChild := false
+		for _, child := range s.Children.members {
```
@lavalamp Do you have an opinion about the O(n^2) loop here? There's no significant different in benchmarks, but I'm wondering why.

jpbetz replied:
Can you binary search s.Children.members instead of scanning? Looks like they are [ordered](https://github.com/kubernetes-sigs/structured-merge-diff/blob/a4e00e9840c22df6cc47c60dcf65984228558387/fieldpath/set.go#L218
). I doubt it will matter for the typical case, but maybe help for the worst case?

jpbetz replied:
I.e. just do `s.Children.members.Descend(child.pathElement)`

apelisse replied:
I can't quite tell if Descend is the right thing, but search is as easy as something like this
```
loc := sort.Search(len(s.members), func(i int) bool {
		return !s.members[i].pathElement.Less(pe)
	})
```

apelisse replied:
But even that is `O(n * log p)` with `n` the number of members and `p` the number of children. Moving the two pointers should be `O(n + p)`

kevindelgado replied:
This results in an almost identical benchmark to both the nested for loop and O(n) implementation from [here](https://github.com/kevindelgado/structured-merge-diff/blob/double-pointer-leaves/fieldpath/set.go#L209-L246).

You can see the code I tested this with [here](https://github.com/kevindelgado/structured-merge-diff/blob/binary-search-leaves/fieldpath/set.go#L209-L231) (in case it's not optimal or what you expected).

The (almost) identical benchmark is:
```
BenchmarkFieldSet/leaves-20-16                            319598              3893 ns/op            2250 B/op         41 allocs/op
BenchmarkFieldSet/leaves-50-16                             81356             14551 ns/op            6851 B/op        140 allocs/op
BenchmarkFieldSet/leaves-100-16                            21072             58362 ns/op           29149 B/op        632 allocs/op
BenchmarkFieldSet/leaves-500-16                             3243            365927 ns/op          146657 B/op       3215 allocs/op
BenchmarkFieldSet/leaves-1000-16                            1417            844380 ns/op          325139 B/op       7190 allocs/op
```

It seems like the random set generator might not ever be testing the worst case, unless there is some under the hood optimizations going on that make the O(n), O(n log n) and O(n^2) approaches all actually do the same thing.

kevindelgado replied:
In summary,

benchmark results:
O(n) == O(n log n) == O(n^2)

so I went with O(n^2) because for readability because:
O(n^2) > [O(n log n)](https://github.com/kevindelgado/structured-merge-diff/blob/binary-search-leaves/fieldpath/set.go#L209-L231) > [O(n)](https://github.com/kevindelgado/structured-merge-diff/blob/double-pointer-leaves/fieldpath/set.go#L209-L246)

I can look into why the random set generation used by the benchmark test is not testing the worst case, but if we're short on time and need to land this asap, we might be better just picking one for now and optimizing later?

kevindelgado replied:
@jpbetz @apelisse thoughts^

apelisse replied:
I'll defer to @lavalamp on this.

apelisse replied:
This file uses the double pointer logic about everywhere (and for good reasons), so I'd still go with that if possible. I've tried to write it, it's not that bad.

kevindelgado replied:
done, ptal @apelisse

apelisse commented on fieldpath/set.go:
```diff
@@ -206,6 +206,29 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
+// Leaves returns a set containing only the leaf paths
+// of a set.
+func (s *Set) Leaves() *Set {
+	members := MakePathElementSet(len(s.Members.members))
```
This one is more likely to over-allocate since members are going to be filtered.

kevindelgado replied:
reverted

lavalamp commented on fieldpath/set.go:
```diff
@@ -206,6 +206,45 @@ func (s *Set) WithPrefix(pe PathElement) *Set {
+		}
+
+		diff := member.Compare(s.Children.members[ic].pathElement)
+		if diff-prevDiff != 0 {
```
This subtraction is really confusing to me, I guess I expected something like:

```go
outer:
for im < len(s.Members.members) {
  member := s.Members.members[im]

  for ic < len(s.Children.members) {
    d := member.Compare(s.Children.members[ic].pathElement)
    if d == 0 {
      ic++
      im++
      continue outer
    } else if d < 0 {
      break
    } else /* if d > 0 */ {
      ic++
    }
  }
  leaves.members = append(leaves.members, member)
}
```

kevindelgado replied:
Ya, that's way cleaner

kevindelgado commented:
All feedback addressed, O(n) Leaves method cleaned up.

PTAL @apelisse

apelisse reviewed (approved):
/lgtm
/approve

apelisse commented:
This adds a new method `Leaves()` in `Set` which is usually not considered a non-backward compatible change. Let's update the minor version.
